presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyExpressions.java

```diff
@@ -212,11 +212,11 @@ public Expression rewriteExpression(Expression node, NodeContext context, Expres
         @Override
         public Expression rewriteLogicalBinaryExpression(LogicalBinaryExpression node, NodeContext context, ExpressionTreeRewriter<NodeContext> treeRewriter)
         {
-            List<Expression> predicates = extractPredicates(node.getType(), node).stream()
-                    .map(expression -> treeRewriter.rewrite(expression, NodeContext.NOT_ROOT_NODE))
-                    .collect(toImmutableList());
+            Expression expression = combinePredicates(node.getType(), extractPredicates(node.getType(), node).stream()
```

Comment:
I think the issue in original code is that it fails to **FULLY** split (extract) the rewritten predicates into parallel predicates with specified logical operator. 

For this case `((A or B) and (A or C)) or D`, the original `predicates` is [`(A or (B and C))`, `(D)`], but the **FULLY** split (extract) result with `OR` should be [`(A)`, `(B and C)`, `(D)`]. Can we just simply do another split (extraction) with the same logical operator after the child expressions are rewritten so that we can leave other code as it is?  
```
List<Expression> predicates = extractPredicates(node.getType(), node).stream()
        .map(expression -> extractPredicates(node.getType(), treeRewriter.rewrite(expression, NodeContext.NOT_ROOT_NODE)))
        .flatMap(expressions -> expressions.stream())
        .collect(toImmutableList());
```